File: Loenn/entities/ss2024/autoReturnFallingBlock.lua
local fakeTilesHelper = require("helpers.fake_tiles")

local fallingBlock = {}

fallingBlock.placements = {}
local directions = {"Up", "Down", "Left", "Right"}

fallingBlock.name = "LylyraHelper/SS2024/AutoReturnFallingBlock"


for _, dir in ipairs(directions) do
    table.insert(fallingBlock.placements, 
        {
        name = "Auto Return Falling Block ("..dir..")",
        data = {
            tiletype = "3",
            climbFall = true,
            behind = false,
            width = 8,
            height = 8,
            resetDelay = 1.0,
            flagOnReset = "",
            flagOnFall = "",
            flagOnLand = "",
            flagTrigger = "",
            resetFlagState = true,
            fallFlagState = true,
            landFlagState = true,
            maxSpeed = 160.0,
            acceleration = 500.0,
            direction = dir,
            landingSound = "",
            returnSound = "",
            shakeSound = "",
            invertFlagTrigger = false,
            returnMaxSpeed = 160.0,
            returnAcceleration = 500.0


            }
        }
    )

end

function fallingBlock.fieldInformation(entity)
	local x = fakeTilesHelper.getFieldInformation("tiletype")
    local y = {
    direction = {
        options = directions,
        editable = false
    }, 
    tiletype = {
        options = fakeTilesHelper.getTilesOptions(layer),
        editable = false
    }}

	return y
end

fallingBlock.sprite = fakeTilesHelper.getEntitySpriteFunction("tiletype", false)

function fallingBlock.depth(room, entity)
    return entity.behind and 5000 or 0
end

return fallingBlock

File: Ahorn/entities/SS2024/autoReturnFallingBlock.jl
module LylyraHelperAutoReturnFallingBlock

using ..Ahorn, Maple

@mapdef Entity "LylyraHelper/SS2024/AutoReturnFallingBlock" AutoReturnFallingBlock(x::Integer, y::Integer, width::Integer=8, height::Integer=8, climbFall::Bool=true, behind::Bool=false,resetDelay::Number=1.0, flagOnReset::String="", flagOnFall::String = "", flagOnLand::String = "", flagTrigger::String = "",resetFlagState::Bool = true,fallFlagState::Bool = true, landFlagState::Bool = true, maxSpeed::Number = 160.0,acceleration::Number = 500.0,direction::String = "Down",landingSound::String = "",returnSound::String = "",shakeSound::String = "",invertFlagTrigger::Bool = true,returnMaxSpeed::Number = 160.0,returnAcceleration::Number = 500.0)



const placements = Ahorn.PlacementDict(
    "Auto Return Falling Block" => Ahorn.EntityPlacement(
        AutoReturnFallingBlock,
        "rectangle",
        Dict{String, Any}(),
        Ahorn.tileEntityFinalizer
    ),
)

Ahorn.editingOptions(entity::AutoReturnFallingBlock) = Dict{String, Any}(
    "tiletype" => Ahorn.tiletypeEditingOptions(),
    "direction" => Maple.move_block_directions
)

Ahorn.minimumSize(entity::AutoReturnFallingBlock) = 8, 8
Ahorn.resizable(entity::AutoReturnFallingBlock) = true, true

Ahorn.selection(entity::AutoReturnFallingBlock) = Ahorn.getEntityRectangle(entity)

Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::AutoReturnFallingBlock, room::Maple.Room) = Ahorn.drawTileEntity(ctx, room, entity)

end